File: platform/macos/monarch_macos/DeviceDefinition.swift
import Foundation

private let _ios : String = "ios";
private let _android : String = "android";

enum TargetPlatform {
    case ios
    case android
}

func targetPlatformFromString(platform: String) -> TargetPlatform {
    return platform == _ios ? .ios : .android;
}

struct DeviceDefinition {
    let id: String
    let name: String
    let logicalResolution: LogicalResolution
    let devicePixelRatio: Double
    let targetPlatform: TargetPlatform

    
    init(id: String, name: String, logicalResolution: LogicalResolution, devicePixelRatio: Double, targetPlatform: TargetPlatform) {
        self.id = id
        self.name = name
        self.logicalResolution = logicalResolution
        self.devicePixelRatio = devicePixelRatio
        self.targetPlatform = targetPlatform
    }
    
    var title: String {
        get {
            let _width = String(format: "%.0f", self.logicalResolution.width)
            let _height = String(format: "%.0f", self.logicalResolution.height)
            return  "\(self.name) | \(_width)x\(_height)"
        }
    }
}

extension DeviceDefinition : InboundChannelArgument {
    
    init(standardMap args: [String : Any]) {
        self.init(
            id: args["id"] as! String,
            name: args["name"] as! String,
            logicalResolution: LogicalResolution.init(standardMap: args["logicalResolution"] as! [String : Any]),
            devicePixelRatio: args["devicePixelRatio"] as! Double,
            targetPlatform: targetPlatformFromString(platform: args["targetPlatform"] as! String)
        )
    }
}

let defaultDeviceDefinition: DeviceDefinition = DeviceDefinition.init(
    id: "ios-iphone-14",
    name: "iPhone 14",
    logicalResolution: LogicalResolution.init(width: 390, height: 844),
    devicePixelRatio: 3.0,
    targetPlatform: .ios)

File: platform/windows/src/device_definition.cpp
#include "device_definition.h"
#include "channels.h"
#include "logical_resolution.h"

TargetPlatform targetPlatformFromString(std::string platform)
{
	return platform == "ios" ? TargetPlatform::ios : TargetPlatform::android;
}

DeviceDefinition::DeviceDefinition()
{
	id = defaultDeviceDefinition.id;
	name = defaultDeviceDefinition.name;
	logicalResolution = defaultDeviceDefinition.logicalResolution;
	devicePixelRatio = defaultDeviceDefinition.devicePixelRatio;
	targetPlatform = defaultDeviceDefinition.targetPlatform;
}

DeviceDefinition::DeviceDefinition(std::string _id, std::string _name,
	LogicalResolution _logicalResolution, double _devicePixelRatio, 
	TargetPlatform _targetPlatform)
{
	id = _id;
	name = _name;
	logicalResolution = _logicalResolution;
	devicePixelRatio = _devicePixelRatio;
	targetPlatform = _targetPlatform;
}

DeviceDefinition::DeviceDefinition(EncodableMap args)
{
	id = std::get<std::string>(MapUtils::getValue(args, "id"));
	name = std::get<std::string>(MapUtils::getValue(args, "name"));
	logicalResolution = LogicalResolution{ std::get<EncodableMap>(MapUtils::getValue(args, "logicalResolution")) };
	devicePixelRatio = std::get<double>(MapUtils::getValue(args, "devicePixelRatio"));
	targetPlatform = targetPlatformFromString(std::get<std::string>(MapUtils::getValue(args, "targetPlatform")));
}

std::string DeviceDefinition::title()
{
	const auto width = std::to_string((int)logicalResolution.width);
	const auto height = std::to_string((int)logicalResolution.height);
	return width + "x" + height + " | " + name;
}

DeviceDefinition defaultDeviceDefinition = DeviceDefinition{
	"ios-iphone-14", 
	"iPhone 14", 
	LogicalResolution{390.0, 844.0},
	3.0,
	TargetPlatform::ios };
